packages/plugin-ext/src/main/browser/debug/plugin-debug-configuration-provider.ts

```diff
@@ -0,0 +1,67 @@
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import {
+    DebugConfigurationProvider,
+    DebugConfigurationProviderDescriptor,
+    DebugConfigurationProviderTriggerKind,
+    DebugExt
+} from '../../../common/plugin-api-rpc';
+import { DebugConfiguration } from '@theia/debug/lib/common/debug-configuration';
+
+export class PluginDebugConfigurationProvider implements DebugConfigurationProvider {
+    private _handle: number;
+    private _type: string;
+    private trigger: DebugConfigurationProviderTriggerKind;
+    provideDebugConfigurations: (folder: string | undefined) => Promise<DebugConfiguration[]>;
+    resolveDebugConfiguration: (folder: string | undefined, debugConfiguration: DebugConfiguration) => Promise<DebugConfiguration | undefined>;
+    resolveDebugConfigurationWithSubstitutedVariables: (folder: string | undefined, debugConfiguration: DebugConfiguration) => Promise<DebugConfiguration | undefined>;
+
+    constructor(description: DebugConfigurationProviderDescriptor,
+        protected readonly debugExt: DebugExt) {
+        this._handle = description.handle;
+        this._type = description.type;
+        this.trigger = description.trigger;
+
+        if (description.provideDebugConfiguration) {
+            this.provideDebugConfigurations = async (folder: string | undefined) => this.debugExt.$provideDebugConfigurationsByHandle(this._handle, folder);
+        }
+
+        if (description.resolveDebugConfigurations) {
+            this.resolveDebugConfiguration =
+                async (folder: string | undefined, debugConfiguration: DebugConfiguration) =>
+                    this.debugExt.$resolveDebugConfigurationByHandle(this._handle, folder, debugConfiguration);
+        }
+
+        if (description.resolveDebugConfigurationWithSubstitutedVariables) {
+            this.resolveDebugConfigurationWithSubstitutedVariables =
+                async (folder: string | undefined, debugConfiguration: DebugConfiguration) =>
+                    this.debugExt.$resolveDebugConfigurationWithSubstitutedVariablesByHandle(this._handle, folder, debugConfiguration);
+        }
+    }
+
+    get handle(): number {
+        return this._handle;
+    }
+
+    get type(): string {
+        return this._type;
+    }
+
+    get triggerKind(): DebugConfigurationProviderTriggerKind {
+        return this.trigger;
+    }
```

Comment:
Since those fields are declared as `readonly` in `interface DebugConfigurationProvider` we can just convert them to `public` fields in this implementation. Code sites dealing with this class as a `DebugConfigurationProvider` won't be able to edit the fields because of the typings.